backend/src/api/Program.cs

```diff
@@ -1,94 +1,108 @@
-using api.Data.DAO;
 using api.Data;
-using api.Services.Tokens;
 using api.Services.Users;
 using Microsoft.AspNetCore.Authentication.JwtBearer;
-using Microsoft.AspNetCore.Identity;
 using Microsoft.IdentityModel.Tokens;
 using Microsoft.OpenApi.Models;
-using System.Text;
 using Microsoft.EntityFrameworkCore;
+using Serilog;
+using api;
+using System.Security.Cryptography;
 
-var builder = WebApplication.CreateBuilder(args);
+const string logFormat =  "[{Timestamp:HH:mm:ss} {Level:u3}] {CorelationId} | {Message:lj}{NewLine}{Exception}";
+Log.Logger = new LoggerConfiguration().Enrich.WithCorrelationId()
+                                             .WriteTo
+                                             .Console(outputTemplate: logFormat)
+                                             .CreateLogger();
 
-builder.Services.AddDbContext<UsersDbContext>(options => options.UseNpgsql(builder.Configuration.GetConnectionString("Users")));
-
-var jwtOptions = new JwtOptions();
-builder.Configuration.Bind("Jwt", jwtOptions);
-if(jwtOptions.Secret is null || jwtOptions.Issuer is null || jwtOptions.Audience is null)
+try
 {
-    throw new Exception("Can't start application without JWT options");
-}
+    var builder = WebApplication.CreateBuilder(args);
+    builder.Services.AddHttpContextAccessor();
+    builder.Host.UseSerilog();
+    builder.Services.AddDbContext<UsersDbContext>(options => options.UseNpgsql(builder.Configuration.GetConnectionString("Users")));
+    var keycloakJwtOptions = new KeycloakJwtOptions();
+    builder.Configuration.Bind("KeycloakJwt", keycloakJwtOptions);
+    if(keycloakJwtOptions.Secret is null || keycloakJwtOptions.Issuer is null || keycloakJwtOptions.Audience is null)
+    {
+        throw new Exception("Can't start application without JWT options");
+    }
 
-var tokenValidationParameters = new TokenValidationParameters()
-{
-    ValidateIssuerSigningKey = true,
-    IssuerSigningKey = new SymmetricSecurityKey(Encoding.ASCII.GetBytes(jwtOptions.Secret)),
-    ValidateAudience = true,
-    ValidAudience = jwtOptions.Audience,
-    ValidateIssuer = true,
-    ValidIssuer = jwtOptions.Issuer,
-    ValidateLifetime = true
-};
-builder.Services.AddSingleton(tokenValidationParameters);
-builder.Services.AddSingleton(jwtOptions);
+    // Create RSA key for offline validation of Keycloak token
+    RSA rsa = RSA.Create();
+    rsa.ImportSubjectPublicKeyInfo(Convert.FromBase64String(keycloakJwtOptions.Secret), out _);
+    var rsaKeycloakSecurityKey = new RsaSecurityKey(rsa)
+    {
+        KeyId = Guid.NewGuid().ToString()
```

Comment:
Are We shure it is Ok for this to differ on application restart?

Reply:
Yes, this value does not matter. We just use this generated key to check if the Keycloak token we got is valid and for that purpose, we just need to have a public key set. The value of KeyId is not important in this process.